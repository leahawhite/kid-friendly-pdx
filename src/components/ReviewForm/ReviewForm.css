.review-form {
  padding: 0 20px;
  text-align: center;
  margin: 0 auto 20px auto;
}

.star-rating-select {
  max-height: 50px !important;
}

.star-rating-select label {
  font-size: 20px;
}

.review-form textarea {
  border: 1px solid #00b8e6;
  display: block;
  font-size: 1.1em;
  margin: 10px 0;
  padding: 10px;
  text-align: left;
  width: 100%;
  min-height: 258px;
  border-radius: 4px;
}

.error {
  color: red;
}

.review-btn, .photo-upload-btn {
  border: 1px solid #00b8e6;
  background-color:  #00b8e6;
  color: white;
  font-weight: bold;
  font-size: 18px;
  border-radius: 8px;
  padding: 8px 12px;
  margin: 10px 0;
  width: 200px;
}

.photo-upload-btn {
  margin: 10px 0 0 0;
}

.photo-upload-btn span {
  margin-left: 10px;
}

.button-container {
  text-align: center;
  margin-bottom: 20px;
}

h2 {
  font-size: 20px;
  margin: 0;
}

@media only screen and (min-width: 700px) {
  .review-form {
    width: 80%;
  }
  
  .button-container {
    display: inline-block;
  }

  .review-btn, .photo-upload-btn {
    display: inline-block;
  }

  .review-btn {
    margin-right: 15px;
  }
}
