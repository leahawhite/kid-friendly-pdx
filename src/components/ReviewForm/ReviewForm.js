import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import StarRatingComponent from 'react-star-rating-component'
import UploadButton from '../UploadButton/UploadButton'
import Images from '../Images/Images'
import Spinner from '../Spinner/Spinner'
import PlacesApiService from '../../services/places-api-service'
import ValidationError from '../ValidationError/ValidationError'
import './ReviewForm.css';

export default class ReviewForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      uploading: null,
      images: [],
      captions: [],
      error: null,
      rating: 0,
      text: "",
      ratingValid: false,
      validationMessages: {
        rating: "",
      },
      fireRedirect: false
    }
  }

  onChange = e => {
    e.preventDefault()
    // how to clear error message other than setstate?
    this.setState({ uploading: true, error: null })

    // this doesn't stop images from being sent to server
    if (e.target.files.length > 3) {
      this.setState({
        error: 'Only 3 images can be uploaded at a time'
      })
    }

    const types = ['image/png', 'image/jpeg', 'image/gif']
    const formData = new FormData()
    for (let i = 0; i < e.target.files.length; i++) {
      if (types.every(type => e.target.files[i].type !== type)) {
        this.setState({
          error: `'${e.target.files[i].type}' is not a supported format`
        })
      }
      formData.append('file', e.target.files[i])
    }

    PlacesApiService.postImagesToCloudinary(formData)
      .then(images => {
        this.setState({ 
          uploading: false,
          images
        })
      })
      .catch(res => {
        this.setState({ error: res.error })
      })
  }

  filter = id => {
    return this.state.images.filter(image => image.public_id !== id)
  }

  removeImage = id => {
    this.setState({
      images: this.state.images.filter(image => image.id !== id)
    })
  }

  addCaption = e => {
    let captions = [...this.state.captions]
    let index = e.target.attributes.getNamedItem('data-index').value
    this.setState({ captions })
    console.log('e.target.value',e.target.value, index)
    console.log('e.target.value[index]', e.target.value[index])
    // console.log('e.target.index', e.target.index)
    // this.setState({
    //   captions: [...this.state.captions, e.target.value[index]]
    // })
    console.log('captions', this.state.captions)
  }

  onError = id => {
    this.setState({ images: this.filter(id) })
  }

  handleSubmit = e => {
    e.preventDefault()
    console.log('images', this.state.images)
    console.log('captions', this.state.captions)
    console.log('e.target', e.target)
    const { place } = this.props
    const { images } = this.state
    const newReview = {
      rating: this.state.rating,
      text: e.target.text.value,
      place_id: place.id
    }
    const newImages = images.map((image, index) => (
      {
        id: image.id,
        src: image.src,
        place_id: place.id,
        title: this.state.captions[index]
      }
    ))

    Promise.all([PlacesApiService.postReview(newReview), PlacesApiService.postImages(newImages)])
    .then(([placesRes, imagesRes]) => {
      if(!placesRes.ok)
        return placesRes.json().then(e => Promise.reject(e))
      if(!imagesRes.ok)
        return imagesRes.json().then(e => Promise.reject(e))
        return Promise.all([placesRes.json(), imagesRes.json()])
    })
    .then(([newReview, newImages]) => {
      // not sure what to do with these
      newReview = {}
      newImages=[]
    })
    .catch(res => {
      this.setState({ error: res.error })
    })
    this.setState({
      fireRedirect: true,
    })
  }
  
  updateRating = (rating) => {
    this.setState({rating}, () => {this.validateRating(rating)})
    console.log(rating)
  }

  validateRating(fieldValue) {
    console.log('validate rating has run')
    const fieldErrors = {...this.state.validationMessages}
    let hasError = false;
    if(fieldValue.length === 0) {
      fieldErrors.rating = 'Star rating is required.';
      hasError = true;
    } else {
      if(fieldValue.length < 1 || fieldValue.length > 5) {
        fieldErrors.rating = 'Please enter a star rating between 1 and 5';
        hasError = true; 
      } else {
        fieldErrors.rating = Number(fieldValue);
        hasError = false;
      }
    }
    this.setState({
      validationMessages: fieldErrors,
      rating: Number(fieldValue),
      ratingValid: !hasError
    }, this.formValid);
  }

  formValid() {
    this.setState({
      formValid: this.state.ratingValid
    });
  }

  renderImages = () => {
    const { images, uploading, captions } = this.state
    if (uploading)
      return <Spinner />
    else if (images.length)
      return <Images 
                images={images} 
                removeImage={this.removeImage} 
                addCaption={this.addCaption}
                onError={this.onError}
                captions={captions}
              />
    else return null;
  }
 
  render() {
    const { place } = this.props
    const { rating, error } = this.state
    const starStyle = {
      display: 'inline',
      fontSize: '34px',
      padding: '0 0 0 10px',
      margin: '-1px 0 0 0',
      verticalAlign: 'middle',
      height: '100%'
    }
    
    return (
      <>
        <div role="alert">
          {error && <p className="error">{error}</p>}
        </div>
        <form className="review-form" onSubmit={this.handleSubmit} method="post" encType="multipart/form-data">
          <div className="star-rating-select">
            <label htmlFor="rating">Your rating: </label>
            <div className="star-container" style={starStyle}>
              <StarRatingComponent 
                name="rating"
                starCount={5}
                value={rating}
                onStarClick={this.updateRating}/>
            </div>
          </div>
          <ValidationError hasError={!this.state.ratingValid} message={this.state.validationMessages.rating}/>
          <div className="text">
            <textarea className="review-text" required placeholder="Share your experience with others. How kid-friendly is this place?" name="text"></textarea>
          </div>
          <div className='upload-container'>
            <UploadButton iconClass="icon" uploadSpan="Upload Photos" onChange={this.onChange} />
            <div className="render-container">
              {this.renderImages()}
            </div>
          </div>
          <div className="button-container">
            <button type="submit" className="review-btn" /*disabled={!this.state.formValid}*/>Post Review</button>
          </div>
        </form>
        {this.state.fireRedirect && (
          <Redirect to={{
            pathname: `/places/${place.id}`,
            state: { place: place }}}
          />
        )}
      </>
    )
  }
}