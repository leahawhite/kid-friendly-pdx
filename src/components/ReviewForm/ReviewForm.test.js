import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom'
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ReviewForm from './ReviewForm';

describe('ReviewForm component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <ReviewForm />
      </BrowserRouter>, 
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders UI as expected', () => {
    const wrapper = shallow(<ReviewForm />)
    expect(toJson(wrapper)).toMatchSnapshot()
  });
})
  
