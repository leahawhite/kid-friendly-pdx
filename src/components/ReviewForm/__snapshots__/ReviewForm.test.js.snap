// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewForm component renders UI as expected 1`] = `
<Fragment>
  <div
    role="alert"
  />
  <form
    className="review-form"
    encType="multipart/form-data"
    method="post"
    onSubmit={[Function]}
  >
    <div
      className="star-rating-select"
    >
      <label
        htmlFor="rating"
      >
        Your rating: 
      </label>
      <div
        className="star-container"
        style={
          Object {
            "display": "inline",
            "fontSize": "34px",
            "height": "100%",
            "margin": "-1px 0 0 0",
            "padding": "0 0 0 10px",
            "verticalAlign": "middle",
          }
        }
      >
        <StarRatingComponent
          editing={true}
          emptyStarColor="#333"
          name="rating"
          onStarClick={[Function]}
          starColor="#ffb400"
          starCount={5}
          value={0}
        />
      </div>
    </div>
    <ValidationError
      hasError={true}
      message=""
    />
    <div
      className="text"
    >
      <textarea
        aria-label="review-text"
        aria-required="true"
        className="review-text"
        name="text"
        placeholder="Share your experience with others. How kid-friendly is this place?"
        required={true}
      />
    </div>
    <_default
      iconClass="icon"
      onChange={[Function]}
      uploadSpan="Upload Photos"
    />
    <div
      className="upload-container"
    >
      <div
        className="render-container"
      />
    </div>
    <div
      className="button-container"
    >
      <button
        aria-label="submit review"
        className="review-btn"
        type="submit"
      >
        Post Review
      </button>
    </div>
  </form>
</Fragment>
`;
